File: Desktop/WinViewer/C#/WinViewer/ViewerForm.cs
using System;
using System.ComponentModel;
using System.IO;
using System.Reflection;
using System.Windows.Forms;
using ActiveReports.Viewer.Helper;
using GrapeCity.ActiveReports.Viewer.Win.Internal.Export;
using ActiveReports.Win.Export;
using GrapeCity.ActiveReports.PageReportModel;
using GrapeCity.ActiveReports.Data;
using LCEArgs = GrapeCity.ActiveReports.Chart.Designer.LocateCredentialsEventArgs;
using GrapeCity.ActiveReports;
using System.Xml;

namespace ActiveReports.Samples.Viewer
{
	using Rom = GrapeCity.ActiveReports.PageReportModel;
	internal partial class ViewerForm : Form
	{
		private ExportForm _exportForm;
		private ExportForm.ReportType? _reportType;
		private string _openFileName;

		public ViewerForm()
		{
			InitializeComponent();
			Icon = Properties.Resource.App;
			SetReportName(null);
		}

		/// <summary>
		/// Loads the document.
		/// </summary>
		/// <param name="file">The file.</param>
		public void LoadDocument(FileInfo file)
		{
			try
			{
				_reportType = ViewerHelper.DetermineReportType(file);
				bool isRdf = ViewerHelper.IsRdf((file));
				var reportServerUri = !isRdf ? ViewerHelper.GetReportServerUri(file) : string.Empty;

				if (!string.IsNullOrEmpty(reportServerUri))
				{

					throw new NotSupportedException("Server report is not supported");
				}
				else
				{
					if (isRdf)
						viewer.LoadDocument(file.FullName);
					else
						switch (_reportType)
						{
							case ExportForm.ReportType.Dashboard:
							case ExportForm.ReportType.PageFpl:
							case ExportForm.ReportType.PageCpl:
								PageReport pageReport = new PageReport(file);
								pageReport.Document.LocateCredentials += OnLocateCredentials;
								viewer.LoadDocument(pageReport.Document);
								break;
							case ExportForm.ReportType.Section:
								SectionReport sectionReport = new SectionReport();
								using (XmlTextReader reader = new XmlTextReader(file.FullName))
								{
									sectionReport.LoadLayout(reader, out System.Collections.ArrayList error);
								}
								sectionReport.Document.Printer.PrinterName = String.Empty;
								sectionReport.LocateCredentials += OnLocateCredentialsSection;
								sectionReport.ResourceLocator = new DefaultResourceLocator { BaseUri = new Uri(file.FullName) };
								viewer.LoadDocument(sectionReport);
								break;
							default:
								throw new ArgumentOutOfRangeException($"Unknown report type to load: \"{_reportType}\"");
						}
				}

				_openFileName = file.FullName;
				exportToolStripMenuItem.Enabled = true;
				SetReportName(openFileDialog.FileName);
			}
			catch (Exception ex)
			{
				viewer.HandleError(ex);
			}
		}

		private void OpenMenuItemHandler(object sender, EventArgs e)
		{
			openFileDialog.FileName = string.Empty;
			if (openFileDialog.ShowDialog(this) == DialogResult.OK)
			{
				LoadDocument(new FileInfo(openFileDialog.FileName));
			}
		}

		private void CloseMenuItemHandler(object sender, EventArgs e)
		{
			Close();
		}

		private void SetReportName(string reportName)
		{
			Text = string.IsNullOrEmpty(reportName) ? string.Format( Properties.Resource.DefaultTitleFormat, Properties.Resource.ViewerFormTitle) : string.Format(Properties.Resource.TitleFormat, Properties.Resource.ViewerFormTitle, Path.GetFileName(reportName));
		}

		private void AboutMenuItemHandler(object sender, EventArgs e)
		{
			const string showAboutBoxMethodName = "ShowAboutBox";
			var attributes = viewer.GetType().GetCustomAttributes(typeof(LicenseProviderAttribute), false);
			if (attributes.Length != 1) return;
			var attr = (LicenseProviderAttribute)attributes[0];
			var provider = ((LicenseProvider)Activator.CreateInstance(attr.LicenseProvider));

			var methodInfo = provider.GetType().GetMethod(showAboutBoxMethodName, BindingFlags.NonPublic | BindingFlags.Instance);

			if (methodInfo != null)
				methodInfo.Invoke(provider, new object[] { viewer.GetType() });
		}

		private void ExportMenuItemHandler(object sender, EventArgs e)
		{
			if (_exportForm == null)
			{
				_exportForm = new ExportForm();
			}

			var reportType = _reportType ?? DetermineOpenedReportType();

			_exportForm.Show(this, new ExportViewer(viewer), reportType);
		}

		private ExportForm.ReportType DetermineOpenedReportType()
		{
			if (viewer.Document != null)
				return ExportForm.ReportType.Section;
			return viewer.OpenedReport == GrapeCity.Viewer.Common.ViewModel.OpenedReport.Fpl
				? ExportForm.ReportType.PageFpl
				: ExportForm.ReportType.PageCpl;
		}

		internal void OnLocateCredentialsSection(object sender, LCEArgs args)
		{
			var pageArgs = ToLocateCredentialsEventArgs(args);

			OnLocateCredentials(sender, pageArgs);

			args.UserName = pageArgs.UserName;
			args.Password = pageArgs.Password;
		}

		private LocateCredentialsEventArgs ToLocateCredentialsEventArgs(LCEArgs args)
		{
			var dataSource = args.DataSource;
			string connectionString = null;
			string dataProvider = null;
			if (dataSource is DbDataSource)
			{
				connectionString = ((DbDataSource)dataSource).ConnectionString;
				dataProvider = ((DbDataSource)dataSource).ProviderName;
			}
			var pageDataSource = new DataSource()
			{
				Name = args.Name,
				ConnectionProperties = new ConnectionProperties
				{
					Prompt = args.PromptText,
					ConnectString = connectionString,
					DataProvider = dataProvider
				},
			};

			return new LocateCredentialsEventArgs(pageDataSource, args.ReportPath, args.PromptText);
		}

		internal void OnLocateCredentials(object sender, LocateCredentialsEventArgs args)
		{
			if (LocateCredentials != null)
				LocateCredentials(sender, args);
			else
			{
				ConnectionProperties connectionProperties = args.DataSource.ConnectionProperties;
				if (connectionProperties == null)
					return;
				var dataSourceName = args.DataSource != null ? args.DataSource.Name : string.Empty;
				string prompt = connectionProperties.Prompt;

				var credential = GetCredentials(new GrapeCity.ActiveReports.PageReportModel.DataSource
				{
					Name = dataSourceName,
					ConnectionProperties = new ConnectionProperties
					{
						Prompt = prompt,
						ConnectString = connectionProperties.ConnectString,
						DataProvider = connectionProperties.DataProvider,
						IntegratedSecurity = connectionProperties.IntegratedSecurity
					},
				});

				args.UserName = credential.UserName;
				args.Password = credential.Password;
			}
		}

		/// <summary>
		/// Raises LocateCredentials event. Occurs when credentials is required for data source loading.
		/// </summary>
		internal event LocateCredentialsEventHandler LocateCredentials;

		internal struct Credentials
		{
			/// <summary>
			/// Gets the user name used to connect to data source.
			/// </summary>
			public string UserName { get; private set; }

			/// <summary>
			/// Gets the password used to connect to data source.
			/// </summary>
			public string Password { get; private set; }

			/// <summary>
			/// Initializes new instance of <see cref="Credentials"/> class.
			/// </summary>
			/// <param name="name">The user name used to connect to data source.</param>
			/// <param name="password">The password used to connect to data source.</param>
			public Credentials(string name, string password) : this()
			{
				UserName = name;
				Password = password;
			}
		}

		private Credentials GetCredentials(Rom.DataSource dataSource)
		{
			Credentials? credentials;
			credentials = new Credentials();

			// Prohibit locate credential if locating credential from designer and credential request's count is more than one.
			// It can be happen on locating credential in subreport.
			string prompt = dataSource.ConnectionProperties.Prompt;
			if (string.IsNullOrEmpty(prompt))
				return credentials ?? new Credentials(string.Empty, string.Empty);

			// Prohibit locate credentials if report is blank or data provider is null
			string dataProvider = dataSource.ConnectionProperties.DataProvider;
			if (string.IsNullOrEmpty(dataProvider))
				return credentials ?? new Credentials(string.Empty, string.Empty);

			credentials = RequestCredentials(dataSource, prompt);

			return credentials.Value;
		}

		private Credentials? RequestCredentials(Rom.DataSource dataSource,
			string prompt)
		{
			var credentialsDialog = new CredentialsDialog(dataSource.Name, prompt);
			if (credentialsDialog.ShowDialog() == DialogResult.OK)
			{
				return new Credentials(credentialsDialog.UserName, credentialsDialog.Password);
			}
			else
				return new Credentials(string.Empty, string.Empty);
		}
	}
}

File: Desktop/WinViewer/VB.NET/WinViewer/ViewerForm.vb
Imports GrapeCity.ActiveReports.Viewer.Win.Internal.Export
Imports System.Reflection
Imports System.IO
Imports System.ComponentModel
Imports System.Text
Imports GrapeCity.Viewer.Common.ViewModel
Imports ActiveReports.Samples.Viewer.ActiveReports.Win.Export
Imports ActiveReports.Samples.Viewer.ActiveReports.Viewer.Helper
Imports LCEArgs = GrapeCity.ActiveReports.Chart.Designer.LocateCredentialsEventArgs
Imports GrapeCity.ActiveReports
Imports GrapeCity.ActiveReports.Data
Imports GrapeCity.ActiveReports.PageReportModel
Imports Rom = GrapeCity.ActiveReports.PageReportModel
Imports System.Xml

Partial Friend Class ViewerForm
	Inherits Form
	Private _exportForm As ExportForm
	Private _reportType As System.Nullable(Of ExportForm.ReportType)
	Private _openFileName As String

	Public Sub New()
		Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
		InitializeComponent()
		Icon = My.Resources.App
		SetReportName(Nothing)
	End Sub

	''' <summary>
	''' Loads the document.
	''' </summary>
	''' <param name="file">The file.</param>
	Public Sub LoadDocument(file As FileInfo)
		Try
			_reportType = ViewerHelper.DetermineReportType(file)
			Dim isRdf As Boolean = ViewerHelper.IsRdf((file))
			Dim reportServerUri = If(Not isRdf, ViewerHelper.GetReportServerUri(file), String.Empty)

			If Not String.IsNullOrEmpty(reportServerUri) Then
				Throw New NotSupportedException("Server report is not supported")
			Else
				''viewer.LoadDocument(file.FullName)
				Select Case _reportType
					Case ExportForm.ReportType.Dashboard, ExportForm.ReportType.PageFpl, ExportForm.ReportType.PageCpl
						Dim pageReport As New PageReport(file)
						AddHandler pageReport.Document.LocateCredentials, AddressOf OnLocateCredentials
						viewer.LoadDocument(pageReport.Document)

					Case ExportForm.ReportType.Section
						Dim sectionReport As New SectionReport()
						Using reader As New XmlTextReader(file.FullName)
							sectionReport.LoadLayout(reader, New System.Collections.ArrayList())
						End Using
						sectionReport.Document.Printer.PrinterName = String.Empty
						AddHandler sectionReport.LocateCredentials, AddressOf OnLocateCredentialsSection
						sectionReport.ResourceLocator = New DefaultResourceLocator() With {
							.BaseUri = New Uri(file.FullName)
						}
						viewer.LoadDocument(sectionReport)
					Case Else
						Throw New ArgumentOutOfRangeException($"Unknown report type to load: ""{_reportType}""")
				End Select

			End If
			_openFileName = file.FullName

			ExportToolStripMenuItem.Enabled = True
			SetReportName(openFileDialog.FileName)
		Catch ex As Exception
			viewer.HandleError(ex)
		End Try
	End Sub

	Private Sub OpenMenuItemHandler(sender As Object, e As EventArgs) Handles OpenToolStripMenuItem.Click
		openFileDialog.FileName = String.Empty
		openFileDialog.Filter = My.Resources.Filter

		If openFileDialog.ShowDialog(Me) = DialogResult.OK Then
			LoadDocument(New FileInfo(openFileDialog.FileName))
		End If
	End Sub

	Private Sub CloseMenuItemHandler(sender As Object, e As EventArgs) Handles ExitToolStripMenuItem.Click
		Close()
	End Sub

	Private Sub SetReportName(reportName As String)
		Text = If(String.IsNullOrEmpty(reportName), String.Format(My.Resources.DefaultTitleFormat, My.Resources.ViewerFormTitle), String.Format(My.Resources.TitleFormat, My.Resources.ViewerFormTitle, Path.GetFileName(reportName)))
	End Sub

	Private Sub AboutMenuItemHandler(sender As Object, e As EventArgs) Handles AboutToolStripMenuItem.Click
		Const showAboutBoxMethodName As String = "ShowAboutBox"
		Dim attributes = viewer.[GetType]().GetCustomAttributes(GetType(LicenseProviderAttribute), False)
		If attributes.Length <> 1 Then
			Return
		End If
		Dim attr As LicenseProviderAttribute = CType(attributes(0), LicenseProviderAttribute)
		Dim provider As Object = (CType(Activator.CreateInstance(attr.LicenseProvider), LicenseProvider))

		Dim methodInfo As Object = provider.[GetType]().GetMethod(showAboutBoxMethodName, BindingFlags.NonPublic Or BindingFlags.Instance)

		If methodInfo IsNot Nothing Then
			methodInfo.Invoke(provider, New Object() {viewer.[GetType]()})
		End If
	End Sub

	Private Sub ExportMenuItemHandler(sender As Object, e As EventArgs) Handles ExportToolStripMenuItem.Click
		If _exportForm Is Nothing Then
			_exportForm = New ExportForm()
		End If

		Dim reportType = If(_reportType, DetermineOpenedReportType())

		_exportForm.Show(Me, New ExportViewer(viewer), reportType)
	End Sub

	Private Function DetermineOpenedReportType() As ExportForm.ReportType
		If viewer.Document IsNot Nothing Then
			Return ExportForm.ReportType.Section
		End If
		Return If(viewer.OpenedReport = OpenedReport.Fpl,
			ExportForm.ReportType.PageFpl,
			ExportForm.ReportType.PageCpl)
	End Function

	Private Sub OnLocateCredentialsSection(sender As Object, args As LCEArgs)
		Dim pageArgs = ToLocateCredentialsEventArgs(args)

		OnLocateCredentials(sender, pageArgs)

		args.UserName = pageArgs.UserName
		args.Password = pageArgs.Password
	End Sub

	Private Function ToLocateCredentialsEventArgs(args As LCEArgs) As LocateCredentialsEventArgs
		Dim dataSource = args.DataSource
		Dim connectionString As String = Nothing
		Dim dataProvider As String = Nothing

		If TypeOf dataSource Is DbDataSource Then
			connectionString = DirectCast(dataSource, DbDataSource).ConnectionString
			dataProvider = DirectCast(dataSource, DbDataSource).ProviderName
		End If

		Dim pageDataSource = New DataSource() With {
		.Name = args.Name,
		.ConnectionProperties = New ConnectionProperties With {
			.Prompt = args.PromptText,
			.ConnectString = connectionString,
			.DataProvider = dataProvider
		}
	}

		Return New LocateCredentialsEventArgs(pageDataSource, args.ReportPath, args.PromptText)
	End Function

	Friend Sub OnLocateCredentials(ByVal sender As Object, ByVal args As LocateCredentialsEventArgs)
		If LocateCredentialsEvent IsNot Nothing Then
			RaiseEvent LocateCredentials(sender, args)
		Else
			Dim connectionProperties As ConnectionProperties = args.DataSource.ConnectionProperties
			If connectionProperties Is Nothing Then
				Return
			End If
			Dim dataSourceName = If(args.DataSource IsNot Nothing, args.DataSource.Name, String.Empty)
			Dim prompt As String = connectionProperties.Prompt

			Dim credential = GetCredentials(New GrapeCity.ActiveReports.PageReportModel.DataSource With {
			.Name = dataSourceName,
			.ConnectionProperties = New ConnectionProperties With {
				.Prompt = prompt,
				.ConnectString = connectionProperties.ConnectString,
				.DataProvider = connectionProperties.DataProvider,
				.IntegratedSecurity = connectionProperties.IntegratedSecurity
			}
		})

			args.UserName = credential.UserName
			args.Password = credential.Password
		End If
	End Sub

	Friend Event LocateCredentials As LocateCredentialsEventHandler

	Friend Structure Credentials
		Public Property UserName As String

		Public Property Password As String

		Public Sub New(name As String, password As String)
			Me.UserName = name
			Me.Password = password
		End Sub
	End Structure

	Private Function GetCredentials(dataSource As Rom.DataSource) As Credentials
		Dim credentials As Credentials? = New Credentials()

		Dim prompt As String = dataSource.ConnectionProperties.Prompt
		If String.IsNullOrEmpty(prompt) Then
			Return If(credentials, New Credentials(String.Empty, String.Empty))
		End If

		Dim dataProvider As String = dataSource.ConnectionProperties.DataProvider
		If String.IsNullOrEmpty(dataProvider) Then
			Return If(credentials, New Credentials(String.Empty, String.Empty))
		End If

		credentials = RequestCredentials(dataSource, prompt)
		Return credentials.Value
	End Function

	Private Function RequestCredentials(dataSource As Rom.DataSource, prompt As String) As Credentials?
		Dim credentialsDialog As New CredentialsDialog(dataSource.Name, prompt)

		If credentialsDialog.ShowDialog() = DialogResult.OK Then
			Return New Credentials(credentialsDialog.UserName, credentialsDialog.Password)
		Else
			Return New Credentials(String.Empty, String.Empty)
		End If
	End Function

End Class
